mtt/src/mtt_types.rs

```diff
@@ -118,3 +116,226 @@ impl MttL1EntryType {
+                    "fracture_1gb_range() error: start: {:x}, current: {:x} first_entry: {first_entry_type:?} current_entry: {mtt_entry:?}",
+                    gb_range_start.bits(), range_addr.bits()
+                )
+            }
+        }
+    }
+}
+
+// Changes the L2 type of 64MB range containing the address to 4Kx16KB.
+// All pages in the range will retain the original MTT type (non-confidential or confidential).
+fn fracture_64mb_range(phys_addr: SupervisorPageAddr) {
+    let mtt_entry = get_l2_entry(phys_addr);
+    let mtt_l1_page = get_l1_page(phys_addr);
+    let l1_page_addr = mtt_l1_page.bits();
+    match mtt_entry {
+        Mtt2MBPages(value) => {
+            // Mark the entire 64MB as non-confidential at the outset
+            // Safety: mtt_l1_page is a supervisor page address that's owned by us, and is at least 4K in size
+            unsafe {
+                core::ptr::write_bytes(l1_page_addr as *mut u8, 0, PageSize::Size4k as usize);
+            }
+            for i in 0u64..=31 {
+                // 01010101b encoding for 4-pages of confidential memory
+                const L1_MTT_CONFIDENTIAL_16K_ENCODING: u8 = 0x55;
+                // Mark the 2MB sub-range as confidential if applicable (128x16K)
+                if value & (1 << i) == MttMemoryType::Confidential as u64 {
+                    // Safety: mtt_l1_page is a supervisor page address that's owned by us, and is at least 4K in size
+                    // The maximum index is 31*128 from the base and we write 128-bytes at a time
+                    unsafe {
+                        core::ptr::write_bytes(
+                            (l1_page_addr + (i * 128)) as *mut u8,
+                            L1_MTT_CONFIDENTIAL_16K_ENCODING,
+                            128,
+                        );
+                    }
+                }
+            }
+
+            set_l2_entry(phys_addr, MttL2EntryType::mtt_l1_dir(mtt_l1_page));
+        }
+        _ => panic!(
+            "fracture_64mb_range() error: address: {:x}, entry: {mtt_entry:?}",
+            phys_addr.bits()
+        ),
+    }
+}
+
+/// Sets the type of the 4K page corresponding to `phys_addr` to the type specified
+/// in `memory_type`
+pub fn set_4k_page_type(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    match get_l2_entry(phys_addr) {
+        MttConfidential1G | MttNonConfidential1G => {
+            fracture_1gb_range(phys_addr);
+            fracture_64mb_range(phys_addr);
+            update_l1_entry(phys_addr, memory_type);
+        }
+        Mtt2MBPages(_) => {
+            fracture_64mb_range(phys_addr);
+            update_l1_entry(phys_addr, memory_type);
+        }
+        MttL1Dir(_) => {
+            update_l1_entry(phys_addr, memory_type);
+        }
+    }
+}
+
+// Gets the base address of the MTT established by the trusted platform-FW.
+// TODO: Implement platform specific functionality.
+#[cfg(all(target_arch = "riscv64", target_os = "none"))]
```

Comment:
This function is used from places that don't share the cfg guard. Does this commit compile?

Reply:
Believe you are referring to this duplicate definition, right? The second one is intended for use for testing, and it compiles and executes the tests just fine as long as the function isn't defined under `#[cfg(test)]`. It's a little odd, but I noticed that the ecall_send() uses the same pattern.

```
// Testing only function intended to emulate functionality
// to get the MTT base.
#[cfg(not(all(target_arch = "riscv64", target_os = "none")))]
fn get_mtt_base() -> u64 {
    static mut MTT_BASE: [u8; 1024 * 1024 * 8] = [0; 1024 * 1024 * 8];
    // Not safe: Intended for testing purposes only
    let addr = unsafe { &MTT_BASE as *const u8 };
    addr as u64
}
```

Reply:
ecall_send uses that to build on riscv/linux as well as test,, hence the strange pattern. Here, cfg(test) should work.

I was confused because the alternate, matching implementation isn't added until a later commit in your series so this looked out of place without that context.